File: mod/common/src/main/java/com/blockforge/common/house/HousePlan.java
package com.blockforge.common.house;

import java.util.List;

public record HousePlan(
        String housePlanId,
        String name,
        HouseStyle style,
        HouseFootprint footprint,
        HouseDimensions dimensions,
        HouseLayout layout,
        HouseRoof roof,
        HouseOpenings openings,
        HouseMaterials materials,
        HouseConstructionOptions options,
        List<HouseModule> modules,
        List<HouseIssue> issues
) {
    public HousePlan {
        modules = modules == null ? List.of() : List.copyOf(modules);
        issues = issues == null ? List.of() : List.copyOf(issues);
    }

    public enum HouseStyle {
        STARTER_COTTAGE,
        MEDIEVAL_HOUSE,
        FARMHOUSE,
        WATCHTOWER_HOUSE,
        STORAGE_HOUSE,
        WORKSHOP,
        MARKET_HOUSE,
        LONGHOUSE,
        MODERN_BOX,
        CUSTOM
    }

    public enum HouseFootprintShape {
        RECTANGLE,
        L_SHAPE,
        T_SHAPE,
        COURTYARD,
        CUSTOM
    }

    public enum HouseRoofType {
        FLAT,
        GABLE,
        HIP,
        PYRAMID,
        TOWER,
        SHED,
        NONE
    }

    public enum HouseRoomType {
        ENTRY,
        MAIN_ROOM,
        BEDROOM,
        STORAGE,
        WORKSHOP,
        KITCHEN,
        STAIR,
        BALCONY,
        CUSTOM
    }

    public enum HouseModuleType {
        FOUNDATION,
        FLOOR,
        WALL,
        ROOF,
        DOOR,
        WINDOW,
        STAIR,
        CHIMNEY,
        PORCH,
        BALCONY,
        TRIM,
        INTERIOR,
        DECORATION
    }

    public record HouseFootprint(int width, int depth, HouseFootprintShape shape) {
    }

    public record HouseDimensions(int floors, int floorHeight, int totalHeight, int foundationHeight) {
    }

    public record HouseRoof(HouseRoofType type, int overhang, int pitch, String mainBlockId, String trimBlockId) {
    }

    public record HouseLayout(List<HouseRoom> rooms, List<HouseConnection> connections) {
        public HouseLayout {
            rooms = rooms == null ? List.of() : List.copyOf(rooms);
            connections = connections == null ? List.of() : List.copyOf(connections);
        }
    }

    public record HouseRoom(String roomId, String name, HouseRoomType type, int x, int z, int width, int depth, int floor) {
    }

    public record HouseConnection(String fromRoomId, String toRoomId, String type) {
    }

    public record HouseOpenings(List<HouseDoor> doors, List<HouseWindow> windows) {
        public HouseOpenings {
            doors = doors == null ? List.of() : List.copyOf(doors);
            windows = windows == null ? List.of() : List.copyOf(windows);
        }
    }

    public record HouseDoor(int x, int y, int z, String facing, String blockId) {
    }

    public record HouseWindow(int x, int y, int z, int width, int height, String facing, String blockId) {
    }

    public record HouseMaterials(
            String foundationBlock,
            String wallBlock,
            String floorBlock,
            String roofBlock,
            String trimBlock,
            String windowBlock,
            String doorBlock,
            String stairBlock,
            String accentBlock
    ) {
    }

    public record HouseConstructionOptions(
            boolean buildFoundation,
            boolean buildInteriorFloor,
            boolean buildRoof,
            boolean addWindows,
            boolean addDoor,
            boolean addChimney,
            boolean addPorch,
            boolean addStairs,
            boolean hollowInterior,
            boolean survivalFriendly,
            boolean useSymmetry
    ) {
        public static HouseConstructionOptions defaults() {
            return new HouseConstructionOptions(true, true, true, true, true, false, false, true, true, true, true);
        }
    }

    public record HouseModule(
            String moduleId,
            HouseModuleType type,
            int x,
            int y,
            int z,
            int width,
            int height,
            int depth,
            String blockKey
    ) {
        public int volume() {
            return Math.max(0, width) * Math.max(0, height) * Math.max(0, depth);
        }
    }

    public record HouseIssue(String severity, String path, String message, String suggestion) {
    }
}

File: src/lib/house/housePlan.ts
import type { BlockType } from "@/types/blueprint";

export type HouseStyle =
  | "starter_cottage"
  | "medieval_house"
  | "farmhouse"
  | "watchtower_house"
  | "storage_house"
  | "workshop"
  | "market_house"
  | "longhouse";

export type HouseRoofType =
  | "flat"
  | "gable"
  | "hip"
  | "pyramid"
  | "tower"
  | "shed"
  | "none";

export type HouseFootprintShape = "rectangle" | "l_shape" | "t_shape" | "courtyard" | "custom";

export type HouseRoomType =
  | "entry"
  | "main_room"
  | "bedroom"
  | "storage"
  | "workshop"
  | "kitchen"
  | "stair"
  | "balcony"
  | "custom";

export type HouseModuleType =
  | "foundation"
  | "floor"
  | "wall"
  | "roof"
  | "door"
  | "window"
  | "stair"
  | "chimney"
  | "porch"
  | "balcony"
  | "trim"
  | "interior"
  | "decoration";

export type HouseFootprint = {
  width: number;
  depth: number;
  shape: HouseFootprintShape;
};

export type HouseDimensions = {
  floors: number;
  floorHeight: number;
  totalHeight: number;
  foundationHeight: number;
};

export type HouseRoof = {
  type: HouseRoofType;
  overhang: number;
  pitch: number;
  mainBlock: BlockType;
  trimBlock: BlockType;
};

export type HouseRoom = {
  roomId: string;
  name: string;
  type: HouseRoomType;
  x: number;
  z: number;
  width: number;
  depth: number;
  floor: number;
};

export type HouseLayout = {
  rooms: HouseRoom[];
  connections: Array<{ fromRoomId: string; toRoomId: string; type: string }>;
};

export type HouseDoor = {
  x: number;
  y: number;
  z: number;
  facing: "north" | "south" | "east" | "west";
  block: BlockType;
};

export type HouseWindow = {
  x: number;
  y: number;
  z: number;
  width: number;
  height: number;
  facing: "north" | "south" | "east" | "west";
  block: BlockType;
};

export type HouseOpenings = {
  doors: HouseDoor[];
  windows: HouseWindow[];
};

export type HouseMaterials = {
  foundation: BlockType;
  wall: BlockType;
  floor: BlockType;
  roof: BlockType;
  trim: BlockType;
  window: BlockType;
  door: BlockType;
  stair: BlockType;
  accent: BlockType;
};

export type HouseConstructionOptions = {
  buildFoundation: boolean;
  buildInteriorFloor: boolean;
  buildRoof: boolean;
  addWindows: boolean;
  addDoor: boolean;
  addChimney: boolean;
  addPorch: boolean;
  addStairs: boolean;
  hollowInterior: boolean;
  survivalFriendly: boolean;
  useSymmetry: boolean;
};

export type HouseModule = {
  moduleId: string;
  type: HouseModuleType;
  x: number;
  y: number;
  z: number;
  width: number;
  height: number;
  depth: number;
  block: BlockType;
};

export type HouseIssue = {
  severity: "info" | "warning" | "error";
  path: string;
  message: string;
  suggestion: string;
};

export type HousePlan = {
  housePlanId: string;
  name: string;
  style: HouseStyle;
  footprint: HouseFootprint;
  dimensions: HouseDimensions;
  layout: HouseLayout;
  roof: HouseRoof;
  openings: HouseOpenings;
  materials: HouseMaterials;
  options: HouseConstructionOptions;
  modules: HouseModule[];
  issues: HouseIssue[];
};

export type HouseGenerationInput = {
  style: HouseStyle;
  width?: number;
  depth?: number;
  floors?: number;
  roofType?: HouseRoofType;
  materials?: Partial<HouseMaterials>;
  options?: Partial<HouseConstructionOptions>;
};

export const houseStyles: HouseStyle[] = [
  "starter_cottage",
  "medieval_house",
  "farmhouse",
  "workshop",
  "storage_house",
  "watchtower_house",
  "market_house",
  "longhouse"
];

export const defaultHouseOptions: HouseConstructionOptions = {
  buildFoundation: true,
  buildInteriorFloor: true,
  buildRoof: true,
  addWindows: true,
  addDoor: true,
  addChimney: false,
  addPorch: false,
  addStairs: true,
  hollowInterior: true,
  survivalFriendly: true,
  useSymmetry: true
};
